### PR TITLE
[FLINK-23354][blob] Limit the size of ShuffleDescriptors in PermanentBlobCache on TaskExecutor

### PR DESCRIPTION
**Pull request by Thesharing**
## What is the purpose of the change

*This pull request intends to add a limit size for ShuffleDescriptors stored in PermanentBlobCache on TaskExecutors. If the size limits exceeds, blob caches are deleted in LRU order. See FLINK-23354 for more details.*

*This pull request is the descendant of FLINK-23218.*

## Brief change log

  - *Introduce BlobCacheSizeTracker to track the size of ShuffleDescriptors stored in PermanentBlobCache on TaskExecutors.*
  - *Introduce `PermanentBlobService#readFile` interface. In PermanentBlobCache, `readFile` makes sure that the file is fully read in the same write lock as the file is accessed. This avoids the scenario that the path is returned as the file is deleted concurrently by other threads. Also `readFile` will track the blobs with BlobCacheSizeTracker. If the size exceeds the limit, PermanentBlobCache will first attempt to delete the files in LRU order and then add the new blob.*

## Verifying this change

This change added tests and can be verified as follows:

  - *Added unit tests for BlobCacheSizeTracker.*
  - *Extended `BlobCacheCleanupTest#testPermanentBlobDeferredCleanup` for cleaning up the track when the job is released.*
  - *Added `DefaultExecutionGraphDeploymentWithSizeLimitTest` to make sure deployment works well when the size limit exceeds.*
  - *Manually verified the change by running an end-to-end test. The job has two JobVertices. One is source, while the other is sink. The parallelism is 8k and 16k. They are connected with an all-to-all edge. Both streaming mode and batch mode are tested.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (**yes** / no / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/blob/AbstractBlobCache.java [hunk 1]
```diff
@@ -249,6 +278,28 @@ public void close() throws IOException {
...
+        final File localFile =
+                new File(
+                        BlobUtils.getStorageLocationPath(
+                                storageDir.getAbsolutePath(), jobId, blobKey));
+        if (!localFile.delete() && localFile.exists()) {
+            log.warn(
+                    "Failed to delete locally cached BLOB {} at {}",
+                    blobKey,
+                    localFile.getAbsolutePath());
+            return false;
```
Comment (zhuzhurk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/blob/AbstractBlobCache.java:298]: Why do we have a return value but add no handling for it?

File: flink-core/src/main/java/org/apache/flink/configuration/BlobServerOptions.java [hunk 2]
```diff
@@ -118,4 +118,10 @@
             key("blob.client.connect.timeout")
                     .defaultValue(0)
                     .withDescription("The connection timeout in milliseconds for the blob client.");
+
+    /** The maximum size limit of blob cache on TaskExecutor. */
+    public static final ConfigOption<Long> BLOB_CACHE_SIZE_LIMIT =
+            key("blob.cache.size.limit")
+                    .defaultValue(1_024 * 1_024 * 1_024 * 10L)
+                    .withDescription("The maximum size limit of blob cache for each TaskExecutor.");
```
Comment (zhuzhurk) [inline flink-core/src/main/java/org/apache/flink/configuration/BlobServerOptions.java:126]: the description needs clearly state how it affects permanent blob and transient blob size.

File: flink-runtime/src/main/java/org/apache/flink/runtime/blob/AbstractBlobCache.java (hunk 1 above)
Comment (Thesharing) [reply flink-runtime/src/main/java/org/apache/flink/runtime/blob/AbstractBlobCache.java:298]: > Why do we have a return value but add no handling for it?

Since the failure of deleting files is tolerable and the log has been outputted here, we can remove the return value.

File: flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobCacheSizeTracker.java [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+ */
+public class BlobCacheSizeTracker {
+
+    private static final Logger LOG = LoggerFactory.getLogger(BlobCacheSizeTracker.class);
+
+    private static final int INITIAL_SIZE = 10_000;
+
+    private final Object lock = new Object();
+
+    private final long capacity;
```
Comment (zhuzhurk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobCacheSizeTracker.java:53]: name it as `sizeLimit` would be easier for understanding

File: flink-core/src/main/java/org/apache/flink/configuration/BlobServerOptions.java (hunk 2 above)
Comment (Thesharing) [reply flink-core/src/main/java/org/apache/flink/configuration/BlobServerOptions.java]: > the description needs to clearly state how it affects permanent blob and transient blob size limit.

I've updated the descriptions here.

File: flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobCacheSizeTracker.java (hunk 3 above)
Comment (Thesharing) [reply flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobCacheSizeTracker.java]: > name it as `sizeLimit` would be easier for understanding

Thank you for pointing this out. Resolved.

**Comment by Thesharing** [comment]
Hi @zhuzhurk. I've refactor the code based on the lastest idea we've discussed before. Would you mind re-reviewing it if you got any free time? Thank you in advance.

File: flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobCacheSizeTrackerTest.java [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+import static org.junit.Assert.assertTrue;
+
+/** Tests for {@link BlobCacheSizeTracker}. */
+public class BlobCacheSizeTrackerTest extends TestLogger {
+
+    private static final int blobSize = 10_000;
+
+    private static final Random random = new Random();
+
+    private static final TemporaryFolder temporaryFolder = new TemporaryFolder();
```
Comment (zhuzhurk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobCacheSizeTrackerTest.java:63]: Can be simplified to be: 
```
@Rule public final TemporaryFolder temporaryFolder = new TemporaryFolder();
```
And the `setUp()` and `cleanUp ()` will not be needed.

File: flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobCacheSizeTrackerTest.java [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+        temporaryFolder.create();
+    }
+
+    @AfterClass
+    public static void cleanUp() throws IOException {
+        temporaryFolder.delete();
+    }
+
+    @Test
+    public void testSizeLimitTrackerStatusUpdate() {
```
Comment (zhuzhurk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobCacheSizeTrackerTest.java:76]: Can be simplified to be `testUpdate `

File: flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobCacheSizeTrackerTest.java [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+
+        assertThat(
+                blobsToDelete,
+                containsInAnyOrder(
+                        Tuple2.of(jobIds.get(0), blobKeys.get(0)),
+                        Tuple2.of(jobIds.get(3), blobKeys.get(3))));
+    }
+
+    @Test
+    public void testTrackPermanentBlobCache() throws Exception {
```
Comment (zhuzhurk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobCacheSizeTrackerTest.java:100]: It is test of `PermanentBlobCache` and should not be in `BlobCacheSizeTrackerTest`.

File: flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobCacheSizeTrackerTest.java [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+            file[3] = getFile(cache, jobId[3], key[3]);
+
+            assertTrue(file[1].exists());
+            assertTrue(file[3].exists());
+            assertFalse(file[2].exists());
+        }
+    }
+
+    @Test
+    public void testGetPermanentCacheWithSizeLimitConcurrently() throws Exception {
```
Comment (zhuzhurk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobCacheSizeTrackerTest.java:170]: It is test of `PermanentBlobCache` and should not be in `BlobCacheSizeTrackerTest`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobCacheSizeTracker.java [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+        synchronized (lock) {
+            for (BlobKey blobKey : blobKeyByJob.computeIfAbsent(jobId, ignore -> new HashSet<>())) {
+                untrack(jobId, blobKey);
+            }
+            blobKeyByJob.remove(jobId);
+        }
+    }
+
+    @VisibleForTesting
+    Set<BlobKey> getBlobKeysByJobId(JobID jobId) {
```
Comment (zhuzhurk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobCacheSizeTracker.java:170]: Seems this method is never used?

File: flink-core/src/main/java/org/apache/flink/configuration/BlobServerOptions.java (hunk 2 above)
Comment (Thesharing) [reply flink-core/src/main/java/org/apache/flink/configuration/BlobServerOptions.java]: This option has been removed.

File: flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobCacheSizeTracker.java [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+            }
+            Long size = caches.remove(key);
+            if (size != null && size >= 0) {
+                total -= size;
+            }
+        }
+    }
+
+    /** Remove the BLOB from the tracker. */
+    public void untrack(@Nullable JobID jobId, BlobKey blobKey) {
```
[the remark this answers is not in the document]
Comment (zhuzhurk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobCacheSizeTracker.java:139]: It can be private

File: flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobCacheSizeTracker.java [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+    public void update(@Nullable JobID jobId, BlobKey blobKey) {
+        checkNotNull(blobKey);
+
+        synchronized (lock) {
+            caches.get(Tuple2.of(jobId, blobKey));
+        }
+    }
+
+    /** Unregister all the tracked BLOBs related to given job. */
+    public void unregisterJob(JobID jobId) {
```
Comment (zhuzhurk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobCacheSizeTracker.java:158]: I prefer the name to be `untrackAll` for easier understanding and also because there is no `registerJob()`

File: flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobCacheSizeTracker.java [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+        }
+    }
+
+    /** Register the target file to the tracker. */
+    public void track(@Nullable JobID jobId, BlobKey blobKey, long size) {
+        checkNotNull(blobKey);
+        checkArgument(size >= 0);
+
+        synchronized (lock) {
+            caches.put(Tuple2.of(jobId, blobKey), size);
```
Comment (zhuzhurk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobCacheSizeTracker.java:107]: `put` -> `putIfAbsent` and increate total only if it is newly added

**Review by Thesharing — COMMENTED** [review]
Thank you so much for your review, @zhuzhurk. You've provided such great suggestions.

File: flink-runtime/src/main/java/org/apache/flink/runtime/blob/PermanentBlobService.java [hunk 12]
```diff
@@ -45,4 +46,21 @@
      * @throws IOException if any other error occurs when retrieving the file
      */
     File getFile(JobID jobId, PermanentBlobKey key) throws IOException;
+
+    /**
+     * Directly read the file associated with the provided job ID and blob key.
+     *
+     * <p>Compared to {@code getFile}, {@code readFile} makes sure that the file is fully read when
```
[the remark this answers is not in the document]
Comment (zhuzhurk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/blob/PermanentBlobService.java]: It's better to illustrate here that why we need to add this interface.

[the remark this answers is not in the document]
Comment (Thesharing) [reply flink-runtime/src/main/java/org/apache/flink/runtime/blob/PermanentBlobService.java]: > It's better to illustrate here that why we need to add this interface.

More detailed comments are added.

File: flink-runtime/src/main/java/org/apache/flink/runtime/blob/PermanentBlobService.java [hunk 13]
```diff
@@ -50,13 +50,20 @@
     /**
      * Returns the content of the file for the BLOB with the provided job ID the blob key.
      *
+     * <p>Compared to {@code getFile}, {@code readFile} will attempt to read the entire file after
+     * retrieving it. If file reading and file retrieving is done in the same WRITE lock, it can
+     * avoid the scenario that the path to the file is deleted concurrently by other threads at the
+     * same time as the file is retrieved and read.
```
Comment (zhuzhurk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/blob/PermanentBlobService.java:56]: maybe "at the same time as the file is retrieved and read" -> "when the file is retrieved but not read yet"?

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/DefaultExecutionGraphDeploymentWithSizeLimitTest.java [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+    @After
+    @Override
+    public void shutdownBlobServer() throws IOException {
+        if (blobServer != null) {
+            blobServer.close();
+        }
+    }
+
+    @Test
+    public void testDeployTasksWithMinimumSizeLimit() throws Exception {
```
[the remark this answers is not in the document]
Comment (zhuzhurk) [reply flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/DefaultExecutionGraphDeploymentWithSizeLimitTest.java]: Ok. Then how about changing the test name to `testDeployMultipleTasksWithBlobCacheSizeLimit`?

File: flink-runtime/src/main/java/org/apache/flink/runtime/blob/PermanentBlobCache.java [hunk 15]
```diff
@@ -187,6 +211,139 @@ public File getFile(JobID jobId, PermanentBlobKey key) throws IOException {
...
+            throws IOException {
+
+        // Check the size limit and delete the files that exceeds the limit
+        final long sizeOfIncomingFile = incomingFile.length();
+        final List<Tuple2<JobID, BlobKey>> blobsToDelete =
+                blobCacheSizeTracker.checkLimit(sizeOfIncomingFile);
+
+        for (Tuple2<JobID, BlobKey> key : blobsToDelete) {
+            deleteFile(key.f0, key.f1);
+            blobCacheSizeTracker.untrack(key);
```
[the remark this answers is not in the document]
Comment (zhuzhurk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/blob/PermanentBlobCache.java]: >>> The only risk is that we will try to delete the blob every time a new blob inserts until the deletion is successful

I think it's better to have to expose the problem. And it also enables retry of failed deletions.

File: flink-runtime/src/main/java/org/apache/flink/runtime/blob/PermanentBlobCache.java [hunk 16]
```diff
@@ -187,6 +211,139 @@ public File getFile(JobID jobId, PermanentBlobKey key) throws IOException {
...
+                        incomingFile,
+                        blobKey,
+                        jobId);
+            }
+        }
+    }
+
+    private void checkLimitAndMoveFile(
+            File incomingFile,
+            @Nullable JobID jobId,
```
[the remark this answers is not in the document]
Comment (zhuzhurk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/blob/PermanentBlobCache.java]: We should also add `checkNotNull` for it.

File: flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobCacheSizeTracker.java [hunk 17]
```diff
@@ -99,44 +101,56 @@ public BlobCacheSizeTracker(long sizeLimit) {
...
+                    LOG.warn(
+                            "The overall size of ShuffleDescriptors in PermanentBlobCache exceeds "
+                                    + "the limit. Limit = [{}], Current: [{}], "
+                                    + "The size of next ShuffleDescriptors: [{}].",
+                            sizeLimit,
+                            total,
+                            size);
+                }
+            } else {
                 LOG.warn(
```
Comment (zhuzhurk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobCacheSizeTracker.java:124]: Is it possible to happen frequently. If yes, I prefer to make it a debug log.

File: flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobCacheSizeTracker.java [hunk 18]
```diff
@@ -99,44 +101,56 @@ public BlobCacheSizeTracker(long sizeLimit) {
...
-            }
-            total += size;
-            if (total > sizeLimit) {
+
+                total += size;
+                if (total > sizeLimit) {
+                    LOG.warn(
+                            "The overall size of ShuffleDescriptors in PermanentBlobCache exceeds "
+                                    + "the limit. Limit = [{}], Current: [{}], "
+                                    + "The size of next ShuffleDescriptors: [{}].",
```
Comment (zhuzhurk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobCacheSizeTracker.java:118]: `ShuffleDescriptors` -> `blob`

File: flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobCacheSizeTracker.java [hunk 19]
```diff
@@ -99,44 +101,56 @@ public BlobCacheSizeTracker(long sizeLimit) {
...
+            if (caches.putIfAbsent(Tuple2.of(jobId, blobKey), size) == null) {
                 blobKeyByJob.computeIfAbsent(jobId, ignore -> new HashSet<>()).add(blobKey);
-            }
-            total += size;
-            if (total > sizeLimit) {
+
+                total += size;
+                if (total > sizeLimit) {
+                    LOG.warn(
+                            "The overall size of ShuffleDescriptors in PermanentBlobCache exceeds "
```
Comment (zhuzhurk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobCacheSizeTracker.java:116]: `ShuffleDescriptors` -> `blob`

File: flink-runtime/src/main/java/org/apache/flink/runtime/blob/PermanentBlobCache.java [hunk 20]
```diff
@@ -247,6 +404,7 @@ public void run() {
 
                         boolean success = false;
                         try {
+                            blobCacheSizeTracker.unregisterJob(jobId);
```
[the remark this answers is not in the document]
Comment (zhuzhurk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/blob/PermanentBlobCache.java]: Thanks for the explanation. I think you are right.

File: flink-runtime/src/main/java/org/apache/flink/runtime/blob/PermanentBlobCache.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (Thesharing) [reply flink-runtime/src/main/java/org/apache/flink/runtime/blob/PermanentBlobCache.java]: I'm not sure the meaning of exposing the problem. If the deletion is failed, `deleteFile` will output a warning log.

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/DefaultExecutionGraphDeploymentWithSizeLimitTest.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (Thesharing) [reply flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/DefaultExecutionGraphDeploymentWithSizeLimitTest.java]: Sure. Resolved.

File: flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobCacheSizeTracker.java (hunk 19 above)
Comment (Thesharing) [reply flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobCacheSizeTracker.java:116]: Thank you for pointing this out. Resolved.

File: flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobCacheSizeTracker.java (hunk 18 above)
Comment (Thesharing) [reply flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobCacheSizeTracker.java:118]: Resolved.

File: flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobCacheSizeTracker.java (hunk 17 above)
Comment (Thesharing) [reply flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobCacheSizeTracker.java:124]: I think this warning shouldn't happen frequently. In `BlobUtil.moveTempFileToStore`, if the blob key duplicates, a warning will be outputted, too. Furthermore, the previous warning ("The overall size of ShuffleDescriptors in PermanentBlobCache exceeds ...") should be warning, too. The overall size exceeds if and only if the size of blob exceeds 100MiB. In this scenario the user should be warned about it.

File: flink-runtime/src/main/java/org/apache/flink/runtime/blob/PermanentBlobService.java (hunk 13 above)
Comment (Thesharing) [reply flink-runtime/src/main/java/org/apache/flink/runtime/blob/PermanentBlobService.java:56]: Thank you for pointing this out. Resolved.

**Review by Thesharing — COMMENTED** [review]
Thank you so much for re-reviewing my pull request. I've came up with several improvements according to your suggestions.

File: flink-runtime/src/test/java/org/apache/flink/runtime/blob/PermanentBlobCacheSizeLimitTest.java [hunk 21]
```diff
@@ -0,0 +1,251 @@
...
+    public void testTrackSizeLimitAndDeleteExcess() throws Exception {
+
+        final BlobKey.BlobType blobType = BlobKey.BlobType.PERMANENT_BLOB;
+
+        final Configuration config = new Configuration();
+        config.setString(
+                BlobServerOptions.STORAGE_DIRECTORY, temporaryFolder.newFolder().getAbsolutePath());
+
+        // The size limit is the size of 2 blobs
+        int sizeLimit = 2;
```
Comment (zhuzhurk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/blob/PermanentBlobCacheSizeLimitTest.java:74]: This is not a size limit but a blob count limit, maybe `maxAcceptedBlobCount`?

File: flink-runtime/src/test/java/org/apache/flink/runtime/blob/PermanentBlobCacheSizeLimitTest.java [hunk 22]
```diff
@@ -0,0 +1,251 @@
...
+            assertEquals(sizeLimit, exists);
+            assertEquals(concurrency - sizeLimit, nonExists);
+        }
+    }
+
+    private static BlobCacheService initBlobCacheServiceWithSizeLimit(
+            final Configuration blobClientConfig,
+            final BlobView blobView,
+            @Nullable final InetSocketAddress serverAddress,
+            final long sizeListOfBlobCacheSizeTracker)
```
Comment (zhuzhurk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/blob/PermanentBlobCacheSizeLimitTest.java:218]: Do you mean `sizeLimitOfBlobCacheSizeTracker `?

File: flink-runtime/src/test/java/org/apache/flink/runtime/blob/PermanentBlobCacheSizeLimitTest.java [hunk 23]
```diff
@@ -0,0 +1,251 @@
...
+                                config,
+                                new VoidBlobStore(),
+                                new InetSocketAddress("localhost", server.getPort()),
+                                sizeLimit * BLOB_SIZE)) {
+
+            server.start();
+
+            // Initialize the data structures
+            JobID[] jobId = new JobID[4];
+            byte[][] data = new byte[4][BLOB_SIZE];
```
Comment (zhuzhurk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/blob/PermanentBlobCacheSizeLimitTest.java:88]: `BLOB_SIZE` does not make sense here. `data[i]` will the reassigned later.

File: flink-runtime/src/test/java/org/apache/flink/runtime/blob/PermanentBlobCacheSizeLimitTest.java [hunk 24]
```diff
@@ -0,0 +1,251 @@
...
+                        initBlobCacheServiceWithSizeLimit(
+                                config,
+                                new VoidBlobStore(),
+                                new InetSocketAddress("localhost", server.getPort()),
+                                sizeLimit * BLOB_SIZE)) {
+
+            server.start();
+
+            // Initialize the data structures
+            JobID[] jobId = new JobID[4];
```
Comment (zhuzhurk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/blob/PermanentBlobCacheSizeLimitTest.java:87]: Let's create a variable to host this number 4.

File: flink-runtime/src/test/java/org/apache/flink/runtime/blob/PermanentBlobCacheSizeLimitTest.java [hunk 25]
```diff
@@ -0,0 +1,251 @@
...
+            File[] file = new File[4];
+
+            // Prepare the blob data
+            byte[] blobData = new byte[BLOB_SIZE];
+            RANDOM.nextBytes(blobData);
+
+            for (int i = 0; i < 4; i++) {
+                jobId[i] = new JobID();
+                data[i] = Arrays.copyOf(blobData, blobData.length);
+                data[i][i] ^= 1;
```
Comment (zhuzhurk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/blob/PermanentBlobCacheSizeLimitTest.java:99]: Maybe add a comments to say that this is to make contents different from each other for later verification?

File: flink-runtime/src/test/java/org/apache/flink/runtime/blob/PermanentBlobCacheSizeLimitTest.java [hunk 26]
```diff
@@ -0,0 +1,251 @@
...
+                file[i] = getFile(cache, jobId[i], key[i]);
+                assertTrue(file[i].exists());
+            }
+
+            // Since the size limit of the cache is the size of 2 blobs,
+            // the first blob is removed and the second one remains
+            assertFalse(file[0].exists());
+            assertTrue(file[1].exists());
+
+            // Get the second blob once again, make the third blob to be the last recently used
```
Comment (zhuzhurk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/blob/PermanentBlobCacheSizeLimitTest.java:118]: `last` -> `least`

File: flink-runtime/src/test/java/org/apache/flink/runtime/blob/PermanentBlobCacheSizeLimitTest.java [hunk 27]
```diff
@@ -0,0 +1,251 @@
...
+            readFileAndVerifyContent(cache, jobId[1], key[1], data[1]);
+
+            // Then get the fourth blob, make sure the third blob is replaced.
+            // Here we cannot get the first blob again, because for transient blob,
```
Comment (zhuzhurk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/blob/PermanentBlobCacheSizeLimitTest.java:122]: The blobs are not transient here. IIUC, this file deletion is not related to BlobServer, right?

File: flink-runtime/src/test/java/org/apache/flink/runtime/blob/PermanentBlobCacheSizeLimitTest.java [hunk 28]
```diff
@@ -0,0 +1,251 @@
...
+    public void testTrackSizeLimitAndDeleteExcessConcurrently() throws Exception {
+
+        final BlobKey.BlobType blobType = BlobKey.BlobType.PERMANENT_BLOB;
+
+        final Configuration config = new Configuration();
+        config.setString(
+                BlobServerOptions.STORAGE_DIRECTORY, temporaryFolder.newFolder().getAbsolutePath());
+
+        final int concurrency = 3;
+        final ExecutorService executor = Executors.newFixedThreadPool(concurrency);
```
Comment (zhuzhurk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/blob/PermanentBlobCacheSizeLimitTest.java:144]: This kind of executor needs to be properly shutdown after used.

File: flink-runtime/src/test/java/org/apache/flink/runtime/blob/PermanentBlobCacheSizeLimitTest.java [hunk 29]
```diff
@@ -0,0 +1,251 @@
...
+
+            byte[] blobData = new byte[BLOB_SIZE];
+            RANDOM.nextBytes(blobData);
+
+            // Put all the BLOBs into the blob server one by one
+            for (int i = 0; i < concurrency; i++) {
+                jobId[i] = new JobID();
+
+                key[i] = put(server, jobId[i], blobData, blobType);
+                assertNotNull(key[i]);
```
Comment (zhuzhurk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/blob/PermanentBlobCacheSizeLimitTest.java:172]: all `blobData` are the same. Why are we not doing the same as in `testTrackSizeLimitAndDeleteExcess()`?

File: flink-runtime/src/test/java/org/apache/flink/runtime/blob/PermanentBlobCacheSizeLimitTest.java [hunk 30]
```diff
@@ -0,0 +1,251 @@
...
+            JobID[] jobId = new JobID[4];
+            byte[][] data = new byte[4][BLOB_SIZE];
+            BlobKey[] key = new BlobKey[4];
+            File[] file = new File[4];
+
+            // Prepare the blob data
+            byte[] blobData = new byte[BLOB_SIZE];
+            RANDOM.nextBytes(blobData);
+
+            for (int i = 0; i < 4; i++) {
```
Comment (zhuzhurk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/blob/PermanentBlobCacheSizeLimitTest.java:96]: Can we extract the blob initialization into a method to be reused by all tests?

File: flink-runtime/src/main/java/org/apache/flink/runtime/blob/PermanentBlobCache.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (zhuzhurk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/blob/PermanentBlobCache.java]: I mean I think it's not "risk is that we will try to delete the blob every time a new blob inserts until the deletion is successful". It's good to log warnings and retry deletions in this case.

File: flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobCacheSizeTrackerTest.java [hunk 31]
```diff
@@ -0,0 +1,168 @@
...
+    public void testCheckLimit() {
+        List<Tuple2<JobID, BlobKey>> keys = tracker.checkLimit(3L);
+
+        assertEquals(1, keys.size());
+        assertEquals(jobId, keys.get(0).f0);
+        assertEquals(blobKey, keys.get(0).f1);
+    }
+
+    @Test
+    public void testZero() {
```
[the remark this answers is not in the document]
Comment (Thesharing) [reply flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobCacheSizeTrackerTest.java]: I feel so awful for these low-level issues. I could do better but I didn't. I should've examined and got them fixed in the early stage. I'll keep in mind and try to do better next time. Thank you so much for being patient and helping me improve the skill. BTW, this unit test has been improved.

File: flink-runtime/src/test/java/org/apache/flink/runtime/blob/PermanentBlobCacheSizeLimitTest.java (hunk 28 above)
Comment (Thesharing) [reply flink-runtime/src/test/java/org/apache/flink/runtime/blob/PermanentBlobCacheSizeLimitTest.java]: Thank you for pointing this out. I should've notice issues like this. This test has been refactored.

File: flink-runtime/src/test/java/org/apache/flink/runtime/blob/PermanentBlobCacheSizeLimitTest.java [hunk 32]
```diff
@@ -0,0 +1,264 @@
...
+
+    @Test
+    public void testTrackSizeLimitAndDeleteExcessSequentially() throws Exception {
+
+        final Configuration config = new Configuration();
+        config.setString(
+                BlobServerOptions.STORAGE_DIRECTORY, temporaryFolder.newFolder().getAbsolutePath());
+
+        try (BlobServer server = new BlobServer(config, new VoidBlobStore());
+                BlobCacheService cache =
```
Comment (zhuzhurk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/blob/PermanentBlobCacheSizeLimitTest.java:75]: Looks to me the `config` can be a class wide field. 
And seems `initBlobCacheServiceWithSizeLimit()` can be param-less.

File: flink-runtime/src/test/java/org/apache/flink/runtime/blob/PermanentBlobCacheSizeLimitTest.java [hunk 33]
```diff
@@ -0,0 +1,254 @@
...
+
+    private static final Random RANDOM = new Random();
+
+    private static final BlobKey.BlobType BLOB_TYPE = BlobKey.BlobType.PERMANENT_BLOB;
+    private static final int BLOB_SIZE = 10_000;
+    // The size limit is the size of 2 blobs
+    private static final int MAX_NUM_OF_ACCEPTED_BLOBS = 2;
+    private static final int TOTAL_NUM_OF_BLOBS = 3;
+
+    private static final Configuration CONFIG = new Configuration();
```
Comment (zhuzhurk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/blob/PermanentBlobCacheSizeLimitTest.java:66]: I think it's better not to make `CONFIG` and `TEMPORARY_FOLDER` static and shared between tests.
It's safer to initialize them before each test.

Comment (Thesharing) [reply flink-runtime/src/test/java/org/apache/flink/runtime/blob/PermanentBlobCacheSizeLimitTest.java]: Yes, although the possibility of having the same JobID and BlobKey is low, but it still makes this case unstable. Resolved.

**Comment by zhuzhurk** [comment]
@flinkbot run azure

**Review by zhuzhurk — APPROVED** [review]
Thanks for addressing all the comments @Thesharing 
The change now looks good to me. Will merge it once CI gives green.

**Comment by Thesharing** [comment]
> Thanks for addressing all the comments @Thesharing
> The change now looks good to me. Will merge it once CI gives green.

Thank you for your reviewing, @zhuzhurk. It helps me a lot. The CI seems to be passed.
